### PR TITLE
(orchestration) fail build after ALL unit tests have been executed - master

### PR DESCRIPTION
**Pull request by clemensutschig**
ok - the idea is pretty simple .. fail delayed if in `WIP` and only during `assemble` after all unit tests executed /in all repos - rather than fail fast (immediately)

@metmajer - please verify

### CODE REVIEW

File: src/org/ods/orchestration/BuildStage.groovy
Comment (clemensutschig) [inline src/org/ods/orchestration/BuildStage.groovy:85]: @metmajer extend with a Check against all repos .. if there is a failed stage..

Comment (clemensutschig) [reply src/org/ods/orchestration/BuildStage.groovy]: @metmajer  - I added what we talked about - please verify .. I don't have a good testcase at hand

File: src/org/ods/orchestration/BuildStage.groovy
Comment (metmajer) [inline src/org/ods/orchestration/BuildStage.groovy:53]: @clemensutschig I am not quite sure about this one. Consider that the call to `jira.reportTestResultsForComponent` a few lines above invokes this method already. I think this to be superfluous.

**Review by metmajer — CHANGES_REQUESTED** [review]
@clemensutschig contribution is most welcome and, although not ideal, I will not merge this change into `3.x`. This requires some time to share thoughts and do proper testing, and unfortunately, we cannot afford a single day of delay at this point in time.

Just to summarise the approach we had discussed:
- The correct behaviour of the Release Manager in the context of reporting failed tests should not depend on the failure mode of ODS components (exit code of test execution of both `0` and `!= 0` are fine).
- [`MROPipelineUntil.prepareExecutePhaseForRepoNamedJob`](https://github.com/opendevstack/ods-jenkins-shared-library/blob/3.x/src/org/ods/orchestration/util/MROPipelineUtil.groovy#L307) calls `executeODSComponent` within `executeBlockAndFailBuild`. The latter throws `Throwable` when an component's pipeline emits a failure.
- When a Release Manager's `Stage` catches this `Throwable`, currently in [`Stage.execute`](https://github.com/opendevstack/ods-jenkins-shared-library/blob/3.x/src/org/ods/orchestration/Stage.groovy#L41), components which have failures AND failing test results shall report these failures into Jira by calling [`JiraUseCase.reportTestResultsForComponent`](https://github.com/opendevstack/ods-jenkins-shared-library/blob/3.x/src/org/ods/orchestration/usecase/JiraUseCase.groovy#L285).

Good with the approach?

File: src/org/ods/orchestration/BuildStage.groovy
Comment (clemensutschig) [reply src/org/ods/orchestration/BuildStage.groovy:53]: Martin.. the warnbuild is idempotent.. and does Not report anythibg to jira.. it parses the Junit Files .. and if failure sets hasfailing tests

**Comment by clemensutschig** [comment]
> 
> Good with the approach?

And thats whats implemented in this pr.. 

Execute Component will during build Not throw but just set failure and warn.. in the build stage ... the last piece is the failure in Case of either failed Tests or other Component failures

File: src/org/ods/orchestration/BuildStage.groovy
Comment (michaelsauter) [inline src/org/ods/orchestration/BuildStage.groovy:91]:
```suggestion
            util.failBuild("Failing build as repositories contain errors!\nFailed: ${failedRepos}")
```

**Review by michaelsauter — COMMENTED** [review]
In general, looks good to me.

I think it is the right approach to handle this on the orchestration level, and not on the component level.

**Comment by hrcornejo** [comment]
With unit tests failing:
image
The test are updated:
image
The behaviour is also correct with a green suite:
image
Verified over an updated 3.x branch with changes of this PR.

**Review by metmajer — APPROVED** [review]

**Pull request merged**
